flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/slowtaskdetector/ExecutionTimeBasedSlowTaskDetector.java

```diff
@@ -0,0 +1,211 @@
+                configuration.getDouble(SlowTaskDetectorOptions.EXECUTION_TIME_BASELINE_RATIO);
+        this.baselineMultiplier =
+                configuration.getDouble(SlowTaskDetectorOptions.EXECUTION_TIME_BASELINE_MULTIPLIER);
+    }
+
+    @Override
+    public void start(
+            final ExecutionGraph executionGraph,
+            final SlowTaskDetectorListener listener,
+            final ComponentMainThreadExecutor mainThreadExecutor) {
+        scheduleTask(executionGraph, listener, mainThreadExecutor);
+    }
+
+    /** Schedule periodical slow task detection. */
+    private void scheduleTask(
+            final ExecutionGraph executionGraph,
+            final SlowTaskDetectorListener listener,
+            final ComponentMainThreadExecutor mainThreadExecutor) {
+        this.scheduledDetectionFuture =
+                mainThreadExecutor.schedule(
+                        () -> {
+                            listener.notifySlowTasks(findSlowTasks(executionGraph));
+                            scheduleTask(executionGraph, listener, mainThreadExecutor);
+                        },
+                        checkIntervalMillis,
+                        TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Given that the parallelism is N and the ratio is R, define T as the median of the first N*R
+     * finished tasks' execution time. The baseline will be T*M, where M is the multiplier. A task
+     * will be identified as slow if its execution time is longer than the baseline.
+     */
+    @VisibleForTesting
+    Map<ExecutionVertexID, Collection<ExecutionAttemptID>> findSlowTasks(
+            final ExecutionGraph executionGraph) {
+        final long currentTimeMillis = System.currentTimeMillis();
+
+        final Map<ExecutionVertexID, Collection<ExecutionAttemptID>> slowTasks = new HashMap<>();
+
+        final List<ExecutionJobVertex> jobVerticesToCheck = getJobVerticesToCheck(executionGraph);
+
+        for (ExecutionJobVertex ejv : jobVerticesToCheck) {
+            final long baseline = getBaseline(ejv, currentTimeMillis);
+
+            for (ExecutionVertex ev : ejv.getTaskVertices()) {
+                if (ev.getExecutionState().isTerminal()) {
+                    continue;
+                }
+
+                final List<ExecutionAttemptID> slowExecutions =
+                        findExecutionsExceedingBaseline(
+                                ev.getCurrentExecutionAttempts(), baseline, currentTimeMillis);
+
+                if (!slowExecutions.isEmpty()) {
+                    slowTasks.put(ev.getID(), slowExecutions);
+                }
+            }
+        }
+
+        return slowTasks;
+    }
+
+    private List<ExecutionJobVertex> getJobVerticesToCheck(final ExecutionGraph executionGraph) {
+        return IterableUtils.toStream(executionGraph.getVerticesTopologically())
+                .filter(ejv -> ejv.isInitialized())
+                .filter(ejv -> ejv.getAggregateState() != ExecutionState.FINISHED)
```

Comment:
Maybe, We can use `ejv -> !ejv.getAggregateState().isTerminal()` here.

Reply:
isTerminal cannot be used to judge here.
Because an ExecutionVertex in this ExecutionJobVertex maybe is failover now.

Reply:
@6591812 is right. For example, for an ExecutionJobVertex `A` it has 2 ExecutionVertex {A1, A2, A3}. If A1 is FAILED, A2 is RUNNING and A3 is FINISHED, `ejv.getAggregateState()` will be `FAILED` but we may still create speculative execution for A2 if it has slow tasks.

Reply:
Got it, thanks.